pkg/noderesourcetopology/cache/discardreserved.go

```diff
@@ -0,0 +1,98 @@
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+
+	topologyv1alpha2 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha2"
+	listerv1alpha2 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/generated/listers/topology/v1alpha2"
+)
+
+type DiscardReserved struct {
+	rMutex         sync.RWMutex
+	reservationMap map[string]map[string]bool // Key is NodeName, value is PodName : reserved status
+	lister         listerv1alpha2.NodeResourceTopologyLister
+}
+
+func NewDiscardReserved(lister listerv1alpha2.NodeResourceTopologyLister) Interface {
+	return &DiscardReserved{
+		lister:         lister,
+		reservationMap: make(map[string]map[string]bool),
+	}
+}
+
+func (pt *DiscardReserved) GetCachedNRTCopy(nodeName string, _ *corev1.Pod) (*topologyv1alpha2.NodeResourceTopology, error) {
+	//readLock
+	pt.rMutex.RLock()
+	defer pt.rMutex.RUnlock()
+	if t, ok := pt.reservationMap[nodeName]; ok {
+		if len(t) > 0 {
+			return nil, fmt.Errorf("node %s is already reserved by another pod", nodeName)
+		}
+	}
+
+	klog.V(5).InfoS("Lister for nodeResTopoPlugin", "lister", pt.lister)
+	nrt, err := pt.lister.Get(nodeName)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get NodeTopologies from NodeResourceTopologyLister")
+	}
+	return nrt, nil
+}
+
+func (pt *DiscardReserved) NodeMaybeOverReserved(nodeName string, pod *corev1.Pod) {}
+func (pt *DiscardReserved) NodeHasForeignPods(nodeName string, pod *corev1.Pod)    {}
+
+func (pt *DiscardReserved) ReserveNodeResources(nodeName string, pod *corev1.Pod) {
+	klog.V(5).InfoS("nrtcache NRT Reserve", "logID", klog.KObj(pod), "node", nodeName)
+	// write lock
+	pt.rMutex.Lock()
+	defer pt.rMutex.Unlock()
+
+	if pt.reservationMap[nodeName] == nil {
+		pt.reservationMap[nodeName] = make(map[string]bool)
+	}
+	podName := klog.KObj(pod).String()
+	pt.reservationMap[nodeName][podName] = true
+
+	klog.V(5).InfoS("nrtcache NRT", "logID", klog.KObj(pod), "node", nodeName, "reservationMap", pt.reservationMap[nodeName])
+}
+
+func (pt *DiscardReserved) UnreserveNodeResources(nodeName string, pod *corev1.Pod) {
+	klog.V(5).InfoS("nrtcache NRT Unreserve", "logID", klog.KObj(pod), "node", nodeName)
+	pt.removeReservationForNode(nodeName, pod)
+}
+
+// PostBind is invoked to cleanup reservationMap
+func (pt *DiscardReserved) PostBind(nodeName string, pod *corev1.Pod) {
```

Comment:
this is interesting. I think I get it but could you please elaborate (gh comment is fine) why you want to remove the reservation on postbind?

Reply:
note really looks very similar (possibly the same) reason we introduced the resync mechanism (using podfingerprint). You can't really be sure about timings here. The NRT update can come anytime. They will come in order, sure, but there's no guarantee about timing and about the sequence of events considering the scheduler event and the NRT update events.

Reply:
Unless I'm missing something, you are totally correct that a NRT update is expected once `PostBind` triggers, but there's no guarantee about WHEN this will happen. NFD can be crashlooping, network can be slow, we can have any network glitch which prevents the NRT update to be propagated [temporary glitch]. But most notably there's no guarantee with respect the occurrence of new pods to be scheduled.

In yet another different way (but still related): what ensures that a NRT update is received before a new scheduling attempt is performed against this node when a new pod comes (new pod scheduling is of course completely async).

In the case of the overreserve, we have the podfingerprint machinery and we intentionally are using a pessimistic approach.

Reply:
You're right but `overreserve` doesn't cache anything if a POD requests resources from multiple NUMA domains(thanks to pesimistic approach and subtracting from every NUMA domain). And it will fetch data from NRT anyway (and temporary glitches also affects this 😄 ). The idea behind Locking nodes is to minimize Admission Errors not to eliminate them.